grades/api.py

```diff
@@ -0,0 +1,179 @@
+    that can be different per program (run?).
+
+    Args:
+        course_run (CourseRun): a course run model object
+
+    Returns:
+        function: a function to be called to compute the final grade
+    """
+    return _compute_grade_for_fa if course_run.course.program.financial_aid_availability else _compute_grade_for_non_fa
+
+
+def _refresh_cache_final_grade(user):
+    """
+    Refreshes the user edX cache for certificates and current grade.
+
+    Args:
+        user (User): a django User
+
+    Returns:
+        None
+    """
+    # get the credentials for the current user for edX
+    user_social = user.social_auth.get(provider=EdxOrgOAuth2.name)
+    utils.refresh_user_token(user_social)
+    # create an instance of the client to query edX
+    edx_client = EdxApi(user_social.extra_data, settings.EDXORG_BASE_URL)
+    CachedEdxDataApi.update_cached_certificates(user, edx_client)
+    CachedEdxDataApi.update_cached_current_grades(user, edx_client)
+
+
+def get_final_grade(user, course_run):
+    """
+    Public function to compute final grades for the a user in a course run.
+
+    Args:
+        user (User): a django User
+        course_run (CourseRun): a course run model object
+
+    Returns:
+        UserFinalGrade: a namedtuple of (grade, passed,) representing the final grade
+            of the user in the course run and whether she passed it
+    """
+    # pull the cached data for the user
+    user_data = CachedEdxUserData(user, course_run.course.program)
+    run_data = user_data.get_run_data(course_run.edx_course_key)
+
+    # pick the right function
+    final_grade_func = _get_compute_func(course_run)
+    return final_grade_func(run_data)
+
+
+def get_users_final_grade_freeze(course_run):
+    """
+    Public function to extract all the users that need a final grade freeze for a course run.
+    All the users that are enrolled in a course run must have frozen final grade.
+
+    Args:
+        course_run (CourseRun): a course run model object
+
+    Returns:
+        queryset: a queryset of users
+    """
+    # get the list of users enrolled in the course
+    users_in_cache = set(CachedEnrollment.get_cached_users(course_run))
+    # get all the users with already frozen final grade
+    users_already_processed = set(FinalGrade.get_frozen_users(course_run))
+    return User.objects.filter(pk__in=users_in_cache.difference(users_already_processed))
+
+
+def freeze_user_final_grade(user, course_run):
+    """
+    Public function to freeze final grades for the a user in a course run.
+
+    Args:
+        user (User): a django User
+        course_run (CourseRun): a course run model object
+
+    Returns:
+        None
+    """
+    # pylint: disable=bare-except
+    # no need to do anything if the course run is not ready
+    if not course_run.can_freeze_grades:
+        log.info(
+            'the grade for user "%s" course "%s" cannot be frozen yet',
+            user.username, course_run.edx_course_key
+        )
```

Comment:
IMHO we should run the app with a log level at least `WARNING` and go down to `INFO` or `DEBUG` only in special cases. 
I might change it to `debug` in case we want to run with the app set at `INFO`